Encrypt dotenv file inplace more than once should be failsafe.

Oh no It hit me again. After changing the configuration in *.sops.yaml* I usually run `sops -i -d; sops -i -e` on the encrypted files to apply the changes. (Maybe there is a better way to do this ?)

But if I run `sops -i -e` twice on a file, then the file can not be encrypted afterwards and sops returns the error  **MAC mismatch. File has ...**. Any succeeding `sops -i -e` call on a file should abort if the file is already encrypted.

@silentpete  Thanks for testing this. I'll consider to create a script with the proposed update command.

> @r10r which version of sops are you using? Sops 3.8.1 does not seem to show the behavior you claim.

sops v3.8.1-49-g418ea23e3

I use a development version to be able set yaml indent to 2 to avoid re-indentation for our yaml files.

```bash
echo HELLO=world > foo.env
echo "stores:" > .sops.yaml
sops -i -e foo.env 
sops -i -e foo.env 
sops -i -d foo.env 
MAC mismatch. File has 712A6E486181E64F8949DAF476EAD3FF0924F26D69A28DADCA03B52BA9DF573C0D2E704FC17985EF0763DBF98C0FA8C3C81A20CF322FBB8AD57E2E600070E3C6, computed 1ADDF28E3BCC682983D0FD44D7C1DD81C15D24B22904B9858E6E7CE8D3C1F30C5FDE2F06A37041C0102E40F201B61EB4264C941F427F2A75C051A710727BB853
```

I can try to reproduce the issue with the current HEAD if required. 

NOTE `echo stores: > .sops.yaml` is required or I get the following exception 

```
panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                        
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xdfa520]                                                                                                                         
                                                                                                                                                                                               
goroutine 1 [running]:                                                                                                                                                                         
github.com/getsops/sops/v3/cmd/sops/common.newDotenvStore(0xf145e0?)                                                                                                                           
        /workdir/cmd/sops/common/common.go:46                                                                                                                                                  
github.com/getsops/sops/v3/cmd/sops/common.StoreForFormat(...)                                                                                                                                 
        /workdir/cmd/sops/common/common.go:162                                                                                                                                                 
github.com/getsops/sops/v3/cmd/sops/common.DefaultStoreForPathOrFormat(0x10bb462?, {0xc000451520, 0xd}, {0x0?, 0x10c35db?})                                                                    
        /workdir/cmd/sops/common/common.go:177 +0x9d                                                                                                                                           
main.inputStore(0xc000187a20, {0xc000451520, 0xd})                                                                                                                                             
        /workdir/cmd/sops/main.go:1085 +0x65                                                                                                                                                   
main.main.func8(0xc000187a20)                                                                                                                                                                  
        /workdir/cmd/sops/main.go:796 +0x770                                                                                                                                                   
github.com/urfave/cli.HandleAction({0xec9b80?, 0x1125d88?}, 0xc00046d6c0?)                                                                                                                     
        /go/pkg/mod/github.com/urfave/cli@v1.22.14/app.go:524 +0x50
github.com/urfave/cli.(*App).Run(0xc00046d6c0, {0xc0000400c0, 0x4, 0x4})
        /go/pkg/mod/github.com/urfave/cli@v1.22.14/app.go:286 +0x766
main.main()
        /workdir/cmd/sops/main.go:1018 +0x35be
```

--- a/cmd/sops/encrypt.go
+++ b/cmd/sops/encrypt.go
@@ -36,7 +36,8 @@ func (err *fileAlreadyEncryptedError) Error() string {
 
 func (err *fileAlreadyEncryptedError) UserError() string {
 	message := "The file you have provided contains a top-level entry called " +
-		"'sops'. This is generally due to the file already being encrypted. " +
+		"'sops', or for flat file formats top-level entries starting with " +
+		"'sops_'. This is generally due to the file already being encrypted. " +
 		"SOPS uses a top-level entry called 'sops' to store the metadata " +
 		"required to decrypt the file. For this reason, SOPS can not " +
 		"encrypt files that already contain such an entry.\n\n" +
@@ -47,10 +48,8 @@ func (err *fileAlreadyEncryptedError) UserError() string {
 }
 
 func ensureNoMetadata(opts encryptOpts, branch sops.TreeBranch) error {
-	for _, b := range branch {
-		if b.Key == "sops" {
-			return &fileAlreadyEncryptedError{}
-		}
+	if opts.OutputStore.HasSopsTopLevelKey(branch) {
+		return &fileAlreadyEncryptedError{}
 	}
 	return nil
 }

--- a/sops.go
+++ b/sops.go
@@ -567,6 +567,12 @@ type ValueEmitter interface {
 	EmitValue(interface{}) ([]byte, error)
 }
 
+// CheckEncryped is the interface for testing whether a branch contains sops
+// metadata. This is used to check whether a file is already encrypted or not.
+type CheckEncryped interface {
+	HasSopsTopLevelKey(TreeBranch) bool
+}
+
 // Store is used to interact with files, both encrypted and unencrypted.
 type Store interface {
 	EncryptedFileLoader
@@ -574,6 +580,7 @@ type Store interface {
 	EncryptedFileEmitter
 	PlainFileEmitter
 	ValueEmitter
+	CheckEncryped
 }
 
 // MasterKeyCount returns the number of master keys available

--- a/stores/dotenv/store.go
+++ b/stores/dotenv/store.go
@@ -175,3 +175,15 @@ func isComplexValue(v interface{}) bool {
 	}
 	return false
 }
+
+// HasSopsTopLevelKey checks whether a top-level "sops" key exists.
+func (store *Store) HasSopsTopLevelKey(branch sops.TreeBranch) bool {
+	for _, b := range branch {
+		if key, ok := b.Key.(string); ok {
+			if strings.HasPrefix(key, SopsPrefix) {
+				return true
+			}
+		}
+	}
+	return false
+}

--- a/stores/dotenv/store_test.go
+++ b/stores/dotenv/store_test.go
@@ -80,3 +80,20 @@ func TestEmitEncryptedFileStability(t *testing.T) {
 		previous = bytes
 	}
 }
+
+func TestHasSopsTopLevelKey(t *testing.T) {
+	ok := (&Store{}).HasSopsTopLevelKey(sops.TreeBranch{
+		sops.TreeItem{
+			Key:   "sops",
+			Value: "value",
+		},
+	})
+	assert.Equal(t, ok, false)
+	ok = (&Store{}).HasSopsTopLevelKey(sops.TreeBranch{
+		sops.TreeItem{
+			Key:   "sops_",
+			Value: "value",
+		},
+	})
+	assert.Equal(t, ok, true)
+}

--- a/stores/ini/store.go
+++ b/stores/ini/store.go
@@ -274,3 +274,8 @@ func (store *Store) EmitExample() []byte {
 	}
 	return bytes
 }
+
+// HasSopsTopLevelKey checks whether a top-level "sops" key exists.
+func (store *Store) HasSopsTopLevelKey(branch sops.TreeBranch) bool {
+	return stores.HasSopsTopLevelKey(branch)
+}

--- a/stores/json/store.go
+++ b/stores/json/store.go
@@ -357,3 +357,13 @@ func (store *Store) EmitExample() []byte {
 	}
 	return bytes
 }
+
+// HasSopsTopLevelKey checks whether a top-level "sops" key exists.
+func (store *Store) HasSopsTopLevelKey(branch sops.TreeBranch) bool {
+	return stores.HasSopsTopLevelKey(branch)
+}
+
+// HasSopsTopLevelKey checks whether a top-level "sops" key exists.
+func (store *BinaryStore) HasSopsTopLevelKey(branch sops.TreeBranch) bool {
+	return stores.HasSopsTopLevelKey(branch)
+}

--- a/stores/stores.go
+++ b/stores/stores.go
@@ -506,3 +506,13 @@ var ExampleFlatTree = sops.Tree{
 		},
 	},
 }
+
+// HasSopsTopLevelKey returns true if the given branch has a top-level key called "sops".
+func HasSopsTopLevelKey(branch sops.TreeBranch) bool {
+	for _, b := range branch {
+		if b.Key == "sops" {
+			return true
+		}
+	}
+	return false
+}

--- a/stores/yaml/store.go
+++ b/stores/yaml/store.go
@@ -417,3 +417,8 @@ func (store *Store) EmitExample() []byte {
 	}
 	return bytes
 }
+
+// HasSopsTopLevelKey checks whether a top-level "sops" key exists.
+func (store *Store) HasSopsTopLevelKey(branch sops.TreeBranch) bool {
+	return stores.HasSopsTopLevelKey(branch)
+}

--- a/stores/yaml/store_test.go
+++ b/stores/yaml/store_test.go
@@ -381,3 +381,20 @@ func TestIndent1(t *testing.T) {
 	assert.Equal(t, string(INDENT_1_OUT), string(bytes))
 	assert.Equal(t, INDENT_1_OUT, bytes)
 }
+
+func TestHasSopsTopLevelKey(t *testing.T) {
+	ok := (&Store{}).HasSopsTopLevelKey(sops.TreeBranch{
+		sops.TreeItem{
+			Key:   "sops",
+			Value: "value",
+		},
+	})
+	assert.Equal(t, ok, true)
+	ok = (&Store{}).HasSopsTopLevelKey(sops.TreeBranch{
+		sops.TreeItem{
+			Key:   "sops_",
+			Value: "value",
+		},
+	})
+	assert.Equal(t, ok, false)
+}

